File: targets/starcoder-todo/prelude.js
// TODO MVU
const todo_eq = ([d1, s1], [d2, s2]) => {
    return d1 === d2 && s1 === s2;
};
const todo_array_eq = (ta1, ta2) => {
    return ta1.length === ta2.length && ta1.every((el, i) => { return todo_eq(el, ta2[i]); });
};
const model_eq = ([b1, ts1], [b2, ts2]) => {
    return b1 === b2 && todo_array_eq(ts1, ts2);
};
const Model_init = ["", []];
const add = (m) => {
    if (m[0] === "") {
        return m[1];
    }
    else {
        return [...m[1], [m[0], false]];
    }
};
const remove = (index, todos) => {
    const removedTodos = [];
    for (let i = 0; i < todos.length; i++) {
        if (i !== index) {
            removedTodos.push(todos[i]);
        }
    }
    return removedTodos;
    // const removedTodos = todos.filter((_, i) => { i !== index });
    // return removedTodos;
};
const toggle = (index, todos) => {
    const toggledTodos = todos.map((t, i) => {
        if (i === index) {
            return [t[0], !t[1]];
        }
        else {
            return t;
        }
    });
    return toggledTodos;
};
export { model_eq, Model_init, add, remove, toggle };

File: targets/ocaml/todo/prelude.ml
(* TODO MVU PRELUDE *)

(* A todo has a description and a status *)
type todo = string * bool

(* A description input buffer and a todo list *)
type model = string * todo list

type action =
  | AddTodo
  | RemoveTodo of int
  | ToggleTodo of int
  | UpdateBuffer of string

type update = model * action -> model

let todo_eq (((d1, s1), (d2, s2)) : todo * todo) : bool = d1 = d2 && s1 = s2

(* TODO: can't use todo_eq, as the comparator is a 'a -> 'a -> bool, not 'a * 'a -> bool *)
let model_eq (((b1, ts1), (b2, ts2)) : model * model) : bool =
  b1 = b2 && List.equal (fun (d1, s1) (d2, s2) -> d1 = d2 && s1 = s2) ts1 ts2

let model_init : model = ("", [])

let add ((description, todos) : model) : todo list =
  if description = "" then todos else (description, false) :: todos

let remove ((index, todos) : int * todo list) : todo list =
  List.filteri (fun i _ -> i != index) todos

let toggle ((index, todos) : int * todo list) : todo list =
  List.mapi
    (fun i (description, finished) ->
      (description, if i = index then not finished else finished))
    todos
